catalog/security/catalog-security-resourceplugin/src/main/java/org/codice/ddf/catalog/security/ResourceUriPolicy.java

```diff
@@ -0,0 +1,187 @@
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.security;
+
+import java.io.Serializable;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+import org.opengis.filter.Filter;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.Result;
+import ddf.catalog.federation.FederationException;
+import ddf.catalog.filter.FilterBuilder;
+import ddf.catalog.operation.Query;
+import ddf.catalog.operation.QueryRequest;
+import ddf.catalog.operation.QueryResponse;
+import ddf.catalog.operation.ResourceRequest;
+import ddf.catalog.operation.ResourceResponse;
+import ddf.catalog.operation.impl.QueryImpl;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+import ddf.catalog.plugin.PolicyPlugin;
+import ddf.catalog.plugin.PolicyResponse;
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.catalog.plugin.impl.PolicyResponseImpl;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+import ddf.security.permission.Permissions;
+
+/**
+ * Restricts how resource URIs are updated and created.
+ * There are security risk to allowing users to update URIs as well as allowing users to
+ * pass in a URI when the metacard is created.
+ */
+public class ResourceUriPolicy implements PolicyPlugin {
+
+    private String[] createPermissions;
+
+    private String[] updatePermissions;
+
+    private CatalogFramework catalogFramework;
+
+    private FilterBuilder filterBuilder;
+
+    @Override
+    public PolicyResponse processPreCreate(Metacard input, Map<String, Serializable> properties)
+            throws StopProcessingException {
+
+        if (input.getResourceURI() != null && StringUtils.isNotEmpty(input.getResourceURI()
+                .toString())) {
+            return new PolicyResponseImpl(null,
+                    Permissions.parsePermissionsFromString(getCreatePermissions()));
+        }
+
+        return new PolicyResponseImpl();
+    }
+
+    @Override
+    public PolicyResponse processPreUpdate(Metacard input, Map<String, Serializable> properties)
+            throws StopProcessingException {
+
+        if (requiresPermissionsToChangeFrom(input.getResourceURI(),
+                getMetacardFromCatalog(input.getId()).getResourceURI())) {
+            return new PolicyResponseImpl(null,
+                    Permissions.parsePermissionsFromString(getUpdatePermissions()));
+        } else {
+            return new PolicyResponseImpl();
+        }
+    }
+
+    private boolean requiresPermissionsToChangeFrom(URI input, URI catalog) {
+
+        return !uriToString(input).equals(uriToString(catalog));
+    }
+
+    private String uriToString(URI uri) {
+        return uri == null ? "" : uri.toString();
```

Comment:
I did some testing. Here is what I found:

Setting the `resource-uri` attribute to `null` causes an exception:

```
Caused by: java.lang.NullPointerException
	at ddf.catalog.data.impl.AttributeImpl.getValue(AttributeImpl.java:112)[357:catalog-rest-endpoint:2.10.0.SNAPSHOT]
	at ddf.catalog.data.impl.MetacardImpl.<init>(MetacardImpl.java:163)[357:catalog-rest-endpoint:2.10.0.SNAPSHOT]
	at ddf.catalog.source.solr.SolrCatalogProvider.copyMetacard(SolrCatalogProvider.java:533)
	at ddf.catalog.source.solr.SolrCatalogProvider.lambda$create$0(SolrCatalogProvider.java:292)
```

Removing the `resource-uri` from the map of attributes to update removes the `resource-uri` from the Solr index. To test this I put a breakpoint on line 416 of SolrCatalogProvider and removed the `resource-uri` from the metacard's attributes. 

>  {
>         "metacard-tags_txt":["resource",
>           "VALID"],
>         "resource-download-url_txt":["https://localhost:8993/services/catalog/sources/dib.distribution/443903e2e0c94a7d9ee7fa4cb8e959ed?transform=resource"],
>         "metadata_xml":["</intelProduct>"],
>         "id_txt":"443903e2e0c94a7d9ee7fa4cb8e959ed",
>         "checksum-algorithm_txt":["Adler32"],
>         "resource-size_txt":["7423"],
>         "created_tdt":["2016-12-15T18:24:57.367Z"],
>         "metacard.modified_tdt":["2016-12-15T18:34:20.498Z"],
>         "metacard.created_tdt":["2016-12-15T18:24:57.356Z"],
>         "checksum_txt":["6e2cb1b3"],
>         "title_txt":["PemFile.java.bava"],
>         "modified_tdt":["2016-12-15T18:34:20.677Z"],
>         "lux_xml":"A==",
>         "metacard_type_name_txt":["common"],
>         "metacard_type_obj": [],
>         "_version_":1553808126885494784}

The results are that if the `resource-uri` attribute is missing from the update request, or the value of the `resource-uri` attribute is null, the `resource-uri` of the metacard will be altered. Therefore, the user should have the correct privileges to update the `resource-uri` if that attribute is missing or if the attribute's value is null.

The conclusion is that an update missing the `resource-uri` attribute should be treated as an update and require permission. The same is true if the attribute is present, but it's value is null, however, an exception will be generated that prevents the update from being written to the Solr index.